test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java

```diff
@@ -875,7 +878,8 @@ ClusterNode restartedNode(Function<MetaData, MetaData> adaptGlobalMetaData, Func
                 final DiscoveryNode newLocalNode = new DiscoveryNode(localNode.getName(), localNode.getId(),
                     UUIDs.randomBase64UUID(random()), // generated deterministically for repeatable tests
                     address.address().getHostString(), address.getAddress(), address, Collections.emptyMap(),
-                    localNode.isMasterNode() ? DiscoveryNodeRole.BUILT_IN_ROLES : emptySet(), Version.CURRENT);
+                    localNode.isMasterNode() && Node.NODE_MASTER_SETTING.get(nodeSettings)
```

Comment:
I think we should make it so that the settings always properly reflect the roles and use `DiscoveryNode.getRolesFromSettings(nodeSettings)` here.